src/poseidon/poseidon_ref.rs

```diff
@@ -0,0 +1,218 @@
+//! Correct, Naive, reference implementation of Poseidon hash function.
+
+use std::convert::TryInto;
+use ark_ff::PrimeField;
+use crate::poseidon::mds::MdsMatrices;
+use crate::poseidon::PoseidonError;
+use crate::poseidon::round_constant::generate_constants;
+use crate::poseidon::round_numbers::calc_round_numbers;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct PoseidonConstants<F: PrimeField> {
+    pub mds_matrices: MdsMatrices<F>,
+    pub round_constants: Vec<F>,
+    pub domain_tag: F,
```

Comment:
What is the `domain_tag` used for?

Reply:
It is concatenated with hash input as the final hash input.

For example, if arity is 2, we have 2+1 hash inputs.

Reply:
for now, domain tag is just `2^ARITY - 1`

Reply:
Is this in the original spec? I thought you were supposed to zero-pad the input like this:

```
fn hash(x: T, y: T) -> T {
    poseidon_permutation_of_width_3(0, x, y)
}
```